### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Sites): characterization of sheaves using 1-hypercovers

### PR DESCRIPTION
**Pull request by joelriou**
In this PR, given a Grothendieck topology `J` on a category `C`, we define a type `J.OneHypercoverFamily` of families of 1-hypercovers. When `H : J.OneHypercoverFamily`, we define a predicate `H.IsGenerating` which means that any covering sieve contains the sieve generated by the underlying covering of one of the 1-hypercovers in the family. If this holds, we show in `OneHypercoverFamily.isSheaf_iff` that a presheaf `P : Cᵒᵖ ⥤ A` is a sheaf iff for any 1-hypercover `E` in the family, the multifork `E.multifork P` is limit.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+  obtain ⟨i, rfl, hi⟩ := hb.exists
+  dsimp at hi
+  rw [id_comp] at hi
+  subst hi
+  rw [assoc, ← P.map_comp, ← op_comp, ← fac,
+    op_comp, P.map_comp, fac'_assoc]
+  exact F.condition
+    { h₁ := le _ (Sieve.ofArrows_mk _ _ _)
+      h₂ := hf
+      w := fac }
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean:117]:
```suggestion
  cases hb with
  | mk i => 
    rw [assoc, ← P.map_comp, ← op_comp, ← fac,
      op_comp, P.map_comp, fac'_assoc]
    exact F.condition
      { h₁ := le _ (Sieve.ofArrows_mk _ _ _)
        h₂ := hf
        w := fac }
```

File: Mathlib/CategoryTheory/Sites/Sieves.lean [hunk 2]
```diff
@@ -138,6 +138,12 @@ inductive ofArrows {ι : Type*} (Y : ι → C) (f : ∀ i, Y i ⟶ X) : Presieve
   | mk (i : ι) : ofArrows _ _ (f i)
 #align category_theory.presieve.of_arrows CategoryTheory.Presieve.ofArrows
 
+lemma ofArrows.exists {ι : Type*} (Y : ι → C) {f : ∀ i, Y i ⟶ X}
+    {Z : C} {g : Z ⟶ X} (hf : ofArrows Y f g) :
+    ∃ (i : ι) (hZ : Z = Y i), g = eqToHom hZ ≫ f i := by
+  obtain ⟨i⟩ := hf
+  exact ⟨i, rfl, by simp⟩
+
 theorem ofArrows_pUnit : (ofArrows _ fun _ : PUnit => f) = singleton f := by
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Sites/Sieves.lean:147]:
```suggestion
theorem ofArrows_pUnit : (ofArrows _ fun _ : PUnit => f) = singleton f := by
```

File: Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    exact ⟨E.isLimitMultifork ⟨_, hP⟩⟩
+  · intro hP
+    rw [Presheaf.isSheaf_iff_multifork]
+    rintro X ⟨S, hS⟩
+    obtain ⟨E, hE, le⟩ := H.exists_oneHypercover S hS
+    exact ⟨IsSheafIff.isLimit hP hE le⟩
+
+end OneHypercoverFamily
+
+/-- Condition that a topology is generated by 1-hypercovers of size `w`. -/
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean:151]:
```suggestion
/-- Condition that a topology is generated by the 1-hypercovers of size `w`. -/
```

**Review by dagurtomas — COMMENTED** [review]
I suggest removing the lemma `ofArrows.exists` which you only use once, and the proof is shorter without using it. Otherwise LGTM

File: Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean]: Usually, I would use `obtain ⟨i⟩ := hb`, but this occasionnaly make certain variable names unreachable (here `Y` becomes `Y✝`), which was the reason for the `ofArrows.exists` lemma. I am happy to see that `cases` does not have this problem. Here, I have used the syntax `cases' hb with i`.

File: Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean (hunk 3 above)
Comment (dagurtomas) [reply Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean:148]: Doesn't it mean that it's generated by *all* the 1-hypercovers?

**Review by dagurtomas — APPROVED** [review]

File: Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean (hunk 3 above)
Comment (joelriou) [reply Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean:148]: Yes, `IsGeneratedByOneHypercovers` means it is generated by *all the* 1-hypercovers.

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean:148]: This is why I suggest adding "the". But I'm not a native English speaker, and in any case it's not that important.

**Comment by erdOne** [comment]
Thanks Joël and Dagur!
maintainer merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**